### ISSUE TITLE
GPU Masking Issue in Secure Vertical XGBoost

### ISSUE THREAD

**Issue by kabelt** [NVIDIA/NVFlare#3687]
### Background

We're trying to implement TreeSHAP on top of the Secure Vertical XGBoost pipeline but have encountered several critical issues that need clarification:

Environment
  - XGBoost: 2.2.0-dev
  - NVFlare: 2.7.0rc3
  - Python: 3.10.18
  - CPU : Xeon® Gold 6130
  - CUDA: 12.4 (Tesla V100-SXM2-16GB)
  - Plugins: Successfully installed both cuda_paillier and nvflare plugins
  - However **cuda_paillier**: Used for both CPU and GPU modes as it's the only working plugin we could compile, 
  - Dataset: Credit card fraud (30 features split across 3 sites as suggested in the official tutorial)

### Issues/Questions

1. GPU Mode Privacy Breach - Complete Exposure of All Features

When using GPU (`tree_method='gpu_hist' with `use_gpus=True`), we discovered that all parties can see all feature split values,
  completely violating federated learning privacy:

**Exact command for reproduction:**
```bash
# GPU mode (PRIVACY BREACH - 0 NaN values)
NVFLARE_XGB_PLUGIN_NAME=cuda_paillier \
NVFLARE_XGB_PLUGIN_PATH=/tmp/nvflare/plugins/libcuda_paillier.so \
python xgb_fl_job.py \
    --data_root /tmp/nvflare/dataset/xgb_dataset/vertical_xgb_data \
    --data_split_mode vertical \
    --num_rounds 3 \
    --secure \
    --use_gpus  # This sets tree_method='gpu_hist'

# CPU mode (WORKING - 36 NaN values with proper masking)
NVFLARE_XGB_PLUGIN_NAME=cuda_paillier \ #yes we used cuda_paillier for CPU and it surprisingly worked!
NVFLARE_XGB_PLUGIN_PATH=/tmp/nvflare/plugins/libcuda_paillier.so \
python xgb_fl_job.py \
    --data_root /tmp/nvflare/dataset/xgb_dataset/vertical_xgb_data \
    --data_split_mode vertical \
    --num_rounds 3 \
    --secure
    # No --use_gpus flag, uses tree_method='hist'
```

<img width="3585" height="3167" alt="Image" src="https://github.com/user-attachments/assets/697de328-194c-4047-bb78-f46079ece00a" />
<img width="3585" height="3167" alt="Image" src="https://github.com/user-attachments/assets/36ecafc8-a971-431c-885a-2928cd143249" />

This could be a complete privacy breach - GPU mode exposes all features to all parties. Is this an expected behavior or a known issue? Or please correct us if our understanding is wrong.

2. IPCL-Python Circular Dependency

We couldn't properly install IPCL-Python, similar to https://github.com/NVIDIA/NVFlare/issues/3479 due to circular dependency issues as well. However, surprisingly:
-   CPU execution runs with cuda_paillier plugin successfully without libnvflare.so + IPCL
-   Models trained by CPU execution can be saved with proper NaN masking
-   Are libnvflare.so + IPCL actually required for secure training on CPU, or is it optional?

3. Inference Implementation Guidance Needed

For TreeSHAP implementation, we need to understand the inference pipeline:
-   We're trying to use bst.predict() on the trained models, but can't find official inference examples
-   Where exactly is the inference code in the secure VFL pipeline?
-   Is there a recommended approach for implementing custom inference logic while maintaining privacy?

4. Performance Discrepancy

The paper "Secure Federated XGBoost with CUDA-accelerated Homomorphic Encryption via NVIDIA FLARE" claims 4.6-30x speedup with GPU, but we observe:

- **CPU Mode**: 71 seconds (with proper NaN masking, 36 NaN values)
- **GPU Mode**: 62 seconds (NO masking, 0 NaN values, all features exposed)
- **Speedup**: Only 12.7% faster but completely loses privacy protection
-   Is this expected given the environment we use above, or please guide us if you think we are doing something wrong?

Thanks in advance for your support and for developing this awesome framework!

**Comment by ZiyueXu77** [comment]
Thanks for the question! Model output masking with gpu_hist is indeed a limitation of the gpu backend of xgb - while such exposure may not be a privacy issue as significant as the gradient pair leakage that could happen without encryption - others seeing the overall model split value of a unknown feature v.s. leaking the label of every single data sample.   

As for the acceleration, there are two separate GPU execution parts: one for XGB training itself ("gpu_hist" v.s. "hist", part of XGB lib capabilities), and the other for FL-side encryption (CPU plugin "libnvflare" v.s. GPU plugin "cuda_paillier")

The 4.6-30x speedup we talked about is on the FL-side encryption, while the speedup you are experiencing is on XGB training (both with GPU plugin for encryption), that part is provided by DLMC XGBoost, and you can refer to https://developer.nvidia.com/blog/gradient-boosting-decision-trees-xgboost-cuda/ for expected acceleration.

We will compose and provide a full table listing all types of security enhancements and accelerations under different combinations of computation options.

**Comment by kabelt** [comment]
> Thanks for the question! Model output masking with gpu_hist is indeed a limitation of the gpu backend of xgb - while such exposure may not be a privacy issue as significant as the gradient pair leakage that could happen without encryption - others seeing the overall model split value of a unknown feature v.s. leaking the label of every single data sample.
> 
> As for the acceleration, there are two separate GPU execution parts: one for XGB training itself ("gpu_hist" v.s. "hist", part of XGB lib capabilities), and the other for FL-side encryption (CPU plugin "libnvflare" v.s. GPU plugin "cuda_paillier")
> 
> The 4.6-30x speedup we talked about is on the FL-side encryption, while the speedup you are experiencing is on XGB training (both with GPU plugin for encryption), that part is provided by DLMC XGBoost, and you can refer to https://developer.nvidia.com/blog/gradient-boosting-decision-trees-xgboost-cuda/ for expected acceleration.
> 
> We will compose and provide a full table listing all types of security enhancements and accelerations under different combinations of computation options.

Thank you for the clarification on GPU masking limitations and performance metrics! This helps us understand the trade-offs better.

  Could you please also provide guidance on the inference implementation, specifically:

  1. Inference with Secure VFL Models:
  After testing, we found that models trained with --secure (CPU mode with NaN masking) cannot perform inference directly - they return AUC=0.5. We understand this is because NaN nodes require cross-party coordination during inference, but we couldn't find documentation or examples for this.

  - Is there an official example of federated inference with NaN node coordination?
  - How should we implement bst.predict() for secure models where parties need to communicate at NaN splits?
  - For TreeSHAP (pred_contribs=True), does the same coordination mechanism work?
  - Or would you suggest we opt for GPU training instead and accept identical model saving among parties as a limitation, so we can conduct inference using the complete model that all parties similarly have?

  Current Status:
  - 🟢 Secure VFL on CPU works : proper model saving with masking
  - 🔴 Secure VFL inference CPU blocked: Models have NaN masking but seemingly need coordination protocol
  - 🎯 Goal: Implement TreeSHAP on secure models for contribution evaluation

  Any code examples or pointers to the relevant inference implementation would be greatly appreciated. 

  Thank you!  🙏

**Comment by ZiyueXu77** [comment]
Hi @kabelt, here is the updated example how the inference can be done with the masked models:
https://github.com/NVIDIA/NVFlare/pull/3707

It includes both the direct standalone experiment for easy understanding, and also the inference and validation with flare

Essentially, the inference will also need to be run with original distributed parties, considering the "feature ownership" - also, as only the label owner has the true label, it will be the only party who can perform auc calculation.

**Comment by ZiyueXu77** [comment]
Hi @kabelt, FYI, we also added an overall summary table regarding acceleration and privacy feature support under different combinations of xgb backend and encryption plugins in this PR: https://github.com/NVIDIA/NVFlare/pull/3710

**Comment by ZiyueXu77** [comment]
Mark closed with inference and doc PR
